### ISSUE TITLE
Feature request: torch.special.gammainc backward pass support

### ISSUE THREAD

**Issue by zpks** [pytorch/pytorch#67763]
## 🚀 Feature
Support for the backward pass with respect to input. It is not supported according to the documentation, referring the reader to opening an issue on the github.
Source: https://pytorch.org/docs/stable/special.html#torch.special.gammainc

## Motivation
The documentation told me to open this issue.
I want to use the gammainc() function as a step between network output and loss calculation. If the backward pass does not work through gammainc(), the network will not work.

## Alternatives
I need the gammainc() function because I need to calculate the generalized exponential integral E_n. A generalized exponential function, akin to scipy.special.expn, would work as an alternative implementation.

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @mruberry

**Comment by mruberry** [comment]
Thanks for the feature request, @zpks!

**Comment by Lezcano** [comment]
We should be able to do this, as we have the digamma function: https://pytorch.org/docs/stable/special.html#torch.special.digamma

Now, this function will be quite unstable, as it grows **very** fast for real inputs. Same for points near a negative number, as the gamma function has poles there.

Knowing what's the application would help us understanding better how to go about this.

In the meantime, you can implement the derivative using `gamma'(t) = gamma(t) * digamma(t)` so the backward is the `conj` of this (if you care about complex inputs).

**Comment by zpks** [comment]
**I now notice that I made an error in my feature request, having confused the upper with the lower incomplete gamma function (`gammainc()` instead of `gammaincc()`). Both however have the same flag directing the reader to open a feature request for backward bass support. Should I change the title of this request?**

**TL:DR I need to calculate my loss through several functions, bottoming out at `gammaincc()`**

Hi, thank you for this answer! I am unsure about how well I will be able to implement your suggestion, as I have not yet needed to interact with autograd on that level. 
In terms of workaround I found on the Wikipedia page on the incomplete gamma function that the upper/lower incomplete gamma functions can be calculated from the Gamma distribution's cumulative distribution function. That function raises a NotImplementedError when called however, so I am afraid that will not work either.

My end goal is to calculate diffraction of optical metasurfaces with ML using a coupled dipole model as a known operator, starting with a 1d periodic structure (also known as a grating). The coupled dipole model here comes down to approximating matter with a few dipoles and using (dyadic) greens functions to calculate fields and  interactions. In this context I need to sum over all lattice sites in an infinite lattice.
After some math the problematic part of the known operator can be expanded to a rapidly convergent form, given by  [1] ![lintoneq](https://user-images.githubusercontent.com/93658362/141289787-5a6f0c9a-44de-4a55-8d6d-069ae5490622.png) where E_n is the exponential integral equal to scipy.special.expn, which may also be expressed as
![image](https://user-images.githubusercontent.com/93658362/141291450-f9c048ed-944c-457f-8c90-88fd99707600.png)
with gamma the upper incomplete gamma function, equal to torch.special.gammaincc(). 

I need this equation both to calculate the coupling effects of dipoles in their lattice and to calculate the electric field resulting from an incoming photon (=electric field) interacting with the coupled dipoles 
The resulting field is the basis for calculating a diffraction spectrum (i.e. reflectivity depending on frequency and angle of incidence), from which I want to calculate the loss. 

The big advantage of using this model is that one can calculate the diffraction from all angles of incidence and at all frequencies from having just those dipoles; calculating a diffraction spectrum straightforwardly can also be done, but only at a single frequency and angle of incidence (up until now). The disadvantage is the rather convoluted road to compute the loss, which I hope is not problematic for autograd.

[1] C.M. Linton, The Green’s function for the two-dimensional Helmholtz equation in periodic domains (1998)

**Comment by Lezcano** [comment]
What about something like this:
```python
import torch

class Incc(torch.autograd.Function):
    @staticmethod
    def forward(ctx, a, x):
        ctx.save_for_backward(a, x)
        return torch.special.gammaincc(a, x)

    @staticmethod
    def backward(ctx, grad_output):
        a, x = ctx.saved_tensors
        grad_x = grad_output * (-x.pow(a-1) * torch.exp(-(x + torch.special.gammaln(a))))
        return (None, grad_x)


incc = Incc.apply

a = torch.rand(3, 3, dtype=torch.float64)
x = torch.rand(3, 3, dtype=torch.float64, requires_grad=True)
torch.autograd.gradcheck(incc, (a, x))
```

Note that this just allows for computing derivatives wrt the second argument. The derivatives wrt. the first argument are considerably more difficult, and we don't have the means to approximate them atm in PyTorch.

cc @kshitij12345 in case you want to implement it as a C++ kernel.

**Comment by Lezcano** [comment]
Scratch all I said above, this function already supports differentiation wrt. the second variable. Is it differentiation wrt. the first variable that you want?

**Comment by zpks** [comment]
No, I want to differentiate with respect to the second variable. I now notice that the variable I want to differentiate to is `other_i`, not `input_i`, meaning that my feature request, in the way that I need it,  has always been satisfied; I merely misinterpreted the documentation. 
My apologies for the inconvenience, and thank you for your support!
